crates/kas-widgets/src/edit/edit_field.rs

```diff
@@ -461,14 +538,17 @@ mod EditField {
                 return false;
             }
 
-            self.input_handler.stop_selecting();
-            self.current.clear_active();
             self.selection.set_max_len(self.text.str_len());
             cx.redraw(&self);
-            if self.current.is_ime() {
-                self.set_ime_cursor_area(cx);
-            }
             self.set_error_state(cx, false);
+
+            // NOTE: we would call self.cancel_selection_and_restart_ime() if we had an EventCx
+            if self.current == CurrentAction::Selection {
+                self.input_handler.stop_selecting();
+                self.current = CurrentAction::None;
+            } else if self.current.is_ime() {
+                cx.request_timer(self.id(), TIMER_RESTART_IME, Default::default());
+            }
```

Comment:
Key/mouse input calls `self.cancel_selection_and_restart_ime()` to handle on-going actions. `fn set_string` can't do this without an `EventCx`.

Instead we use a timer as a callback to restart IME in case `set_string` is called. Hacky but probably fine.